bin/audit.py

```diff
@@ -0,0 +1,152 @@
+# from gfactoryXmlToDB import *
+from collections import defaultdict
+import yaml
+import requests
+import sys
+import urllib.request
+try:
+    import xml.etree.cElementTree as ET
+except ImportError:
+    import xml.etree.ElementTree as ET
+
+
+treeDump = False
+
+
+def getGfactoryData(gfactoryDB, xml):
+    response = urllib.request.urlopen(xml)
+    gfactoryPage = response.read()
+    root = ET.fromstring(gfactoryPage)
+
+    for entries in root.findall('entries'):
+        for entry in entries.findall('entry'):
+            for attrs in entry.findall('attrs'):
+                for attr in attrs.findall('attr'):
+                    if attr.get('name') == 'GLIDEIN_ResourceName':
+                        if treeDump:
+                            print(attr.get('value'))
+                        # try:
+                        #     gfactoryDB.add(attr.get('value'))
+                        # except KeyError:
+                        #     gfactoryDB = {attr.get('value')}
+                        gfactoryDB.add(attr.get('value'))
+
+
+def getTopologyData(topologyDB):
+    # insert Names under a dictionary that stores 4 "groupname"-set pairs
+    # Structure of the dictionary:
+    # {'resourceGroups': {},
+    #  'facilities': {},
+    #  'sites': {},
+    #  'resources': {}}
+    #
+    # The XML has the following hierarchy: (only showing info we need)
+    # | root
+    # | --ResourceGroup
+    # | ----Facility
+    # | ----Site
+    # | ----Resources
+    # | --------Resource
+    response = urllib.request.urlopen(
+        "https://my.opensciencegrid.org/rgsummary/xml")
+    topologyPage = response.read()
+    topologyRoot = ET.fromstring(topologyPage)
+
+    # -- file version of parsing. Currently this script read from string
+    # topologyTree = ET.parse("resource_topology.xml")
+    # topologyRoot = topologyTree.getroot()
+
+    for child in topologyRoot.findall('ResourceGroup'):
+        # adding resourceGroup Name attribute to a set
+        name = child.find('GroupName')
+        if treeDump:
+            print("| " + name.text)
+        try:
+            topologyDB['resourceGroups'].add(name.text)
+        except KeyError:
+            topologyDB['resourceGroups'] = {name.text}
```

Comment:
This sort of thing is actually what `collections.defaultdict` is designed for!

```
>>> from collections import defaultdict
>>> topologyDB = defaultdict(set)
>>> topologyDB['ResourceGroups'].add('foo')
>>> topologyDB['ResourceGroups']
{'foo'}
```

Reply:
Yeah I agree with this part. I was using {} instead of set() to initialize the variable and had to do this. Will fix it